src/Workspaces/Remote/Core/Diagnostics/PerformanceQueue.cs

```diff
@@ -0,0 +1,226 @@
+            var squareLength = Math.Sqrt(data.Count);
+
+            return (average, stddev / squareLength);
+        }
+
+        private class Snapshot
+        {
+            /// <summary>
+            /// Raw performance data. 
+            /// Keyed by analyzer unique number got from AnalyzerNumberAssigner.
+            /// Value is delta (TimeSpan - minSpan) among span in this snapshot
+            /// </summary>
+            private readonly Dictionary<int, double> _performanceMap;
+
+            public Snapshot(IEnumerable<(string, TimeSpan)> snapshot, int unitCount) :
+                this(Convert(snapshot), unitCount)
+            {
+            }
+
+            public Snapshot(IEnumerable<(int, TimeSpan)> rawData, int unitCount)
+            {
+                _performanceMap = new Dictionary<int, double>();
+
+                Reset(_performanceMap, rawData, unitCount);
+            }
+
+            public void Update(IEnumerable<(string, TimeSpan)> rawData, int unitCount)
+            {
+                Reset(_performanceMap, Convert(rawData), unitCount);
+            }
+
+            public void AppendAnalyzers(HashSet<int> analyzerSet)
+            {
+                analyzerSet.UnionWith(_performanceMap.Keys);
+            }
+
+            public double? GetTimeSpanInMillisecond(int assignedAnalyzerNumber)
+            {
+                if (!_performanceMap.TryGetValue(assignedAnalyzerNumber, out var value))
+                {
+                    return null;
+                }
+
+                return value;
+            }
+
+            private void Reset(
+                Dictionary<int, double> map, IEnumerable<(int assignedAnalyzerNumber, TimeSpan timeSpan)> rawData, int fileCount)
+            {
+                // get smallest timespan in the snapshot
+                var minSpan = rawData.Select(kv => kv.timeSpan).Min();
+
+                // for now, we just clear the map, if reusing dictionary blindly became an issue due to
+                // dictionary grew too big, then we need to do a bit more work to determine such case
+                // and re-create new dictionary
+                map.Clear();
+
+                // map is normalized to current timespan - min timspan of the snapshot
+                foreach (var (assignedAnalyzerNumber, timeSpan) in rawData)
+                {
+                    map[assignedAnalyzerNumber] = (timeSpan.TotalMilliseconds - minSpan.TotalMilliseconds) / fileCount;
+                }
+            }
+
+            private static IEnumerable<(int, TimeSpan)> Convert(IEnumerable<(string analyzerId, TimeSpan timeSpan)> rawData)
+            {
+                return rawData.Select(kv => (AnalyzerNumberAssigner.Instance.GetUniqueNumber(kv.analyzerId), kv.timeSpan));
+            }
+        }
+
+        /// <summary>
+        /// Assign unique number to diagnostic analyzers
+        /// </summary>
+        private class AnalyzerNumberAssigner
```

Comment:
❔ Why not just make this a `static class`?

Reply:
because it has state. generally static type is not supposed to be stateful.